File: src/code/cpp/graph/bfs.cpp
#include <queue>
#include <unordered_set>
#include <vector>

using namespace std;


int fn(vector<vector<int>>& graph) {
    queue<int> que;
    unordered_set<int> seen;
    que.push(START_NODE);
    seen.insert(START_NODE);
    int ans = 0;

    while (!que.empty()) {
        int node = que.front();
        que.pop();
        // TODO: logic
        for (int neighbor: graph[node]) {
            if (seen.find(neighbor) == seen.end()) {
                seen.insert(neighbor);
                que.push(neighbor);
            }
        }
    }
}

File: src/code/ruby/graph/bfs.rb
require 'set'

def fn(graph)
  queue = [START_NODE]
  seen  = Set.new([START_NODE])
  ans   = 0

  until queue.empty?
    node = queue.shift
    # TODO: logic
    graph[node].each do |neighbor|
      unless seen.include?(neighbor)
        seen.add(neighbor)
        queue << neighbor
      end
    end
  end

  ans
end
